From 93548b15379cc59b84cdabcbd1dbbece646a2ffc Mon Sep 17 00:00:00 2001
From: Johan Klokkhammer Helsing <johan.helsing@qt.io>
Date: Fri, 1 Nov 2019 11:24:26 +0100
Subject: [PATCH 03/18] Client: Fix crash when showing a child window with a
 hidden parent

[ChangeLog][QPA plugin] Fixed a crash when showing a window with a hidden
parent.

Now we just avoid creating the subsurface, so nothing is shown. Seems to be
the same behavior as on xcb.

Fixes: QTBUG-79674
Change-Id: Ia46fcd9a0da5aad4704816a41515cb1e128ac65f
Reviewed-by: Paul Olav Tvete <paul.tvete@qt.io>
(cherry picked from commit 962b9be7992cef672cb6307af5653c97382c334f)
Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 src/client/qwaylanddisplay.cpp | 4 ++++
 src/client/qwaylandwindow.cpp  | 7 ++++---
 2 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/src/client/qwaylanddisplay.cpp b/src/client/qwaylanddisplay.cpp
index 86cfe1a0..aef31961 100644
--- a/src/client/qwaylanddisplay.cpp
+++ b/src/client/qwaylanddisplay.cpp
@@ -109,6 +109,10 @@ struct ::wl_region *QWaylandDisplay::createRegion(const QRegion &qregion)
         return NULL;
     }
 
+    // Make sure we don't pass NULL surfaces to libwayland (crashes)
+    Q_ASSERT(parent->object());
+    Q_ASSERT(window->object());
+
     return mSubCompositor->get_subsurface(window->object(), parent->object());
 }
 
diff --git a/src/client/qwaylandwindow.cpp b/src/client/qwaylandwindow.cpp
index dc664304..e7a01a54 100644
--- a/src/client/qwaylandwindow.cpp
+++ b/src/client/qwaylandwindow.cpp
@@ -139,9 +139,10 @@ void QWaylandWindow::initWindow()
     if (shouldCreateSubSurface()) {
         Q_ASSERT(!mSubSurfaceWindow);
 
-        QWaylandWindow *p = static_cast<QWaylandWindow *>(QPlatformWindow::parent());
-        if (::wl_subsurface *ss = mDisplay->createSubSurface(this, p)) {
-            mSubSurfaceWindow = new QWaylandSubSurface(this, p, ss);
+        auto *parent = static_cast<QWaylandWindow *>(QPlatformWindow::parent());
+        if (parent->object()) {
+            if (::wl_subsurface *subsurface = mDisplay->createSubSurface(this, parent))
+                mSubSurfaceWindow = new QWaylandSubSurface(this, parent, subsurface);
         }
     } else if (shouldCreateShellSurface()) {
         Q_ASSERT(!mShellSurface);
-- 
2.20.1

